src/allocator.c

```diff
@@ -654,27 +654,39 @@ _dispatch_allocator_enumerate(task_t remote_task,
 		vm_address_t zone_address, memory_reader_t reader,
 		void (^recorder)(vm_address_t, void *, size_t, bool *stop))
 {
-	const size_t heap_size = remote_dal->dal_magazine_size;
-	const size_t dc_size = remote_dal->dal_allocation_size;
-	const size_t dc_flags_offset = remote_dal->dal_allocation_isa_offset;
-	bool stop = false;
-	void *heap;
-
-	while (zone_address) {
-		// FIXME: improve this by not faulting everything and driving it through
-		//        the bitmap.
-		kern_return_t kr = reader(remote_task, zone_address, heap_size, &heap);
-		size_t offs = remote_dal->dal_first_allocation_offset;
-		if (kr) return kr;
-		while (offs < heap_size) {
-			void *isa = *(void **)(heap + offs + dc_flags_offset);
-			if (isa && isa != remote_dal->dal_deferred_free_isa) {
-				recorder(zone_address + offs, heap + offs, dc_size, &stop);
-				if (stop) return KERN_SUCCESS;
+	if (zone_address)
```

Comment:
This seems like it just changes a while statement into an if branch then a do-while. What is the benefit of this??

Reply:
To avoid the assignments if not needed at all.